README.md
musicxml2lilypond
===========
Python package that converts a makam score in MusicXml format to [LilyPond](http://lilypond.org/)

Introduction
------------
This repository provides a simple and reliable conversion of a makam music score from MusicXML format to LilyPond format.

You can obtain 2200 makam music scores in MusicXML format from [SymbTr](http://dunya.compmusic.upf.edu/makam/) collection. Alternatively, you can use the SymbTr-txt files in the collection and generate the MusicXML files with limited modifications using the [MusicXMLConverter](https://github.com/burakuyar/MusicXMLConverter) package.

The LilyPond format and the LilyPond software itself are a powerful combination for music engraving, with vast use cases. Moreover, when the software convert the LilyPond format to svg, it embeds a precise mapping of the musical elements in the LilyPond file onto the generated vectors in the svg file. In turn, the musical elements in the svg files can be rendered synchronous to the music playback in relevant applications.

Usage
------

```python
from musicxml2lilypond.scoreconverter import ScoreConverter

# input MusicXML file
xml_file = "samplescores/pesendide--sazsemaisi--aksaksemai----iii_selim.xml"

# output lilypond file
ly_file = "samplescores/pesendide--sazsemaisi--aksaksemai----iii_selim.ly"

# output json file storing the mappings beteen ly and musicxml files.
map_file = "samplescores/pesendide--sazsemaisi--aksaksemai----iii_selim.json"

# instantiate
converter = ScoreConverter()
ly_stream, mapping = converter.convert(xml_file, ly_out=ly_file, 
                                       render_metadata=False)
```

Installation
-------------

**Note:** This package is in active development. The capabilities and calls might change significantly until the first stable release. For this reason we suggest the users to stick to the released versions, for the most reliable performance.

If you want to install musicxml2lilypond, it is recommended to install the package and its dependencies into a virtualenv. In the terminal, do the following:

    virtualenv env
    source env/bin/activate
    python setup.py install

If you want to be able to edit files and have the changes be reflected, then
install the repository like this instead:

    pip install -e .

Authors
-------
Hasan Sercan Atlı	hsercanatli AT gmail DOT com  
Burak Uyar	burakuyar AT gmail DOT com  
Sertan Şentürk	contact AT sertansenturk DOT com  
Andrés Ferraro	andres DOT ferraro AT upf DOT edu  

Reference
-------
Sercan's Thesis


Changelog.md
#### musicxml2lilypond v0.9.6
 - Added time signature changes

#### musicxml2lilypond v0.9.5
 - Removed mapping json saving
 - Made the lilypond output more compact
 - \t are replaced by double spaces in the ly file
 - Row and column indices in the mapping are fixed

#### musicxml2lilypond v0.9.4
 - Added simple gracenote parsing
 - Better string maintainability by enforcing ```__future__.unicode_literals```
 - Added ```unittests```
 - Created and modularized ```MusicXMLReader```
 - Cleaned setup requirements
 - Improved ```demo.ipynb```

#### musicxml2lilypond v0.9.3
 - Corrected database path
 - Database path is constructed OS-independently during conversion

#### musicxml2lilypond v0.9.2
 - The xml input can be given directly as a string to the ```ScoreConverter.run``` method
 - Tidier output saving in the ```ScoreConverter.run``` method

#### musicxml2lilypond v0.9.1
 - Refactored the code to improve readability and maintenance

#### musicxml2lilypond v0.9.0
 - First public release


musicxml2lilypond/unittests/conversion_tests.py
from __future__ import unicode_literals
from musicxml2lilypond.scoreconverter import ScoreConverter
import os
import json


def test_instrumental():
    name = "pesendide--sazsemaisi--aksaksemai----iii_selim"
    _converter(name)


def test_vocal():
    name = "ussak--sarki--duyek--aksam_oldu_huzunlendim--semahat_ozdenses"
    _converter(name)


def _converter(symbtr_name):
    xml_file, ly_file, map_file = _get_files_from_symbtrname(symbtr_name)

    converter = ScoreConverter()
    ly_stream, mapping = converter.convert(
        xml_file, ly_out=None, render_metadata=True)

    with open(ly_file, 'r') as f:
        saved_ly = f.read().decode('utf-8')

    assert ly_stream == saved_ly, u'{0:s}.ly file does not match'.format(
        symbtr_name)

    saved_map = json.load(open(map_file, 'r'))
    saved_map = [tuple(m) for m in saved_map]
    assert mapping == saved_map, u'A different mapping for {0:s} is produced.'\
        .format(symbtr_name)


def _get_files_from_symbtrname(symbtr_name):
    # folder
    curr_folder = os.path.dirname(os.path.abspath(__file__))
    file_template = os.path.join(curr_folder, 'data', symbtr_name)

    # files
    xml_file = file_template + '.xml'
    ly_file = file_template + '.ly'
    map_file = file_template + '.json'

    return xml_file, ly_file, map_file


musicxml2lilypond/MusicXMLReader.py
# coding=utf-8
from __future__ import unicode_literals, division
import warnings
import xml.etree.ElementTree as eT
from xml.etree.ElementTree import ParseError

__author__ = 'hsercanatli', 'burakuyar', 'andresferrero', 'sertansenturk'


class MusicXMLReader(object):
    makam_accidentals = {'quarter-flat': '-1',
                         'slash-flat': '-4',
                         'flat': '-5',
                         'double-slash-flat': '-8',
                         'quarter-sharp': '+1',
                         'sharp': '+4',
                         'slash-quarter-sharp': '+5',
                         'slash-sharp': '+8'}

    @classmethod
    def read(cls, xml_in):
        """
        :param xml_in:
        :rtype: object
        """

        # setting the xml tree
        parser = _XMLCommentHandler()
        try:
            try:  # document
                tree = eT.parse(xml_in, parser)
                root = tree.getroot()
            except IOError:  # string input
                root = eT.fromstring(xml_in, parser)
        except ParseError:
            raise ParseError("Line 36(ish): Error parsing MusicXML file.")

        # getting beats and beat type
        beat_type = root.find('part/measure/attributes/time/beat-type').text
        beats = root.find('part/measure/attributes/time/beats').text

        # getting key signatures
        keysig = cls._get_key_signature(root)

        # makam, form and usul information
        makam, form, usul = cls._get_makam_form_usul(root)

        # work title
        work_title = cls._get_title(root)

        # composer and lyricist
        composer, lyricist = cls._get_composer_lyricist(root)

        # reading the xml measure by measure
        measures = cls._get_measures(root)

        return (measures, makam, usul, form, beats, beat_type, keysig,
                work_title, composer, lyricist)

    @staticmethod
    def _get_key_signature(root):
        keysig = {}  # keys: notes, values: type of note accident
        for xx, key in enumerate(
                root.findall('part/measure/attributes/key/key-step')):
            keysig[key.text] = root.findall(
                'part/measure/attributes/key/key-accidental')[xx].text

        return keysig

    @staticmethod
    def _get_makam_form_usul(root):
        if root.find('part/measure/direction/direction-type/words').text:
            # the information is stored in the form below:
            # "Makam: [makam], Form: [form], Usul: [usul] "
            cultural_info = root.find(
                'part/measure/direction/direction-type/words').text

            attributes = []
            for info in cultural_info.split(","):
                attributes.append(''.join(info.split(": ")[1]).strip())

            makam, form, usul = attributes
        else:
            warnings.warn("Makam, Form and Usul information do not exist.")
            makam = form = usul = ''

        return makam, form, usul

    @staticmethod
    def _get_title(root):
        if root.find('work/work-title').text:
            work_title = ''.join(root.find('work/work-title').text).strip()
        else:
            work_title = ''.strip()

        return work_title

    @staticmethod
    def _get_composer_lyricist(root):
        identification = [item.text for item in root.findall(
            'identification/creator')]
        if len(identification) == 2:
            composer = ''.join(identification[0]).strip()
            lyricist = ''.join(identification[1]).strip()
        else:
            composer = ''.join(identification[0]).strip()
            lyricist = ''.strip()
        return composer, lyricist

    @classmethod
    def _get_measures(cls, root):
        # tempo
        bpm = float(root.find('part/measure/direction/sound').attrib['tempo'])
        divisions = float(root.find('part/measure/attributes/divisions').text)
        quarter_note_len = 60000.0 / bpm

        # read measures
        measures = []
        for measure_index, measure in enumerate(root.findall('part/measure')):
            temp_measure = []
            # all notes in the current measure
            for note_index, note in enumerate(measure.findall('note')):
                # symbtr-txt id, which is stored in MusicXML
                symbtr_txt_id = cls._get_symbtr_txt_id(note)

                # rest
                rest = cls._chk_rest(note)

                # pitch and octave information of the current note
                pitch_step, octave = cls._get_pitchstep_octave(note, rest)

                # symbolic duration without considering dotted, tuple etc. The
                # duration info will be handled properly during the LilyPond
                # conversion
                normal_dur = cls._get_normal_dur(note, divisions,
                                                 quarter_note_len)

                # accident inf
                acc = cls._get_accidental(note)

                # dotted or not
                dot = cls._chk_dotted(note)

                # tuplet or not
                tuplet = cls._chk_tuplet(note)

                # lyrics
                lyrics = cls._get_lyrics(note)

                # appending attributes to the temp note
                temp_note = [pitch_step, octave, acc, dot, tuplet, rest,
                             normal_dur, symbtr_txt_id, lyrics]

                temp_measure.append(temp_note)

            # add temp measure to the measure
            measures.append(temp_measure)
        return measures

    @staticmethod
    def _get_pitchstep_octave(note, rest):
        if rest:
            pitch_step = 'r'
            octave = 'r'
        elif note.find('pitch') is not None:  # pitch
            pitch_step = note.find('pitch/step').text.lower()
            octave = note.find('pitch/octave').text
        else:
            raise ValueError("The element should have been a note or "
                             "rest")

        return pitch_step, octave

    @staticmethod
    def _get_symbtr_txt_id(note):
        if note.find("symbtrid").text:
            return int(note.find("symbtrid").text)
        else:
            return None

    @staticmethod
    def _chk_rest(note):
        return note.find('rest') is not None

    @staticmethod
    def _get_normal_dur(note, divisions, quarter_note_len):
        if note.find('duration') is None:  # grace note
            return None
        else:
            dur = note.find('duration').text  # get the true duration
            return (int(quarter_note_len * float(dur) / divisions) /
                    quarter_note_len)

    @classmethod
    def _get_accidental(cls, note):
        if note.find('accidental') is not None:
            return cls.makam_accidentals[note.find('accidental').text]
        else:
            return 0

    @staticmethod
    def _chk_dotted(note):
        return note.find('dot') is not None

    @staticmethod
    def _chk_tuplet(note):
        return note.find('time-modification') is not None

    @staticmethod
    def _get_lyrics(note):
        if note.find('lyric/text').text is not None:
            return note.find('lyric/text').text
        else:
            return ''


class _XMLCommentHandler(eT.XMLTreeBuilder):
    def __init__(self):
        super(_XMLCommentHandler, self).__init__()

        # assumes ElementTree 1.2.X
        self._parser.CommentHandler = self.handle_comment

    def handle_comment(self, data):
        self._target.start("symbtrid", {})
        if data and 'symbtr_txt_note_index' in data:
            data = data.replace('symbtr_txt_note_index ', '')
        self._target.data(data)
        self._target.end("symbtrid")


musicxml2lilypond/ScoreConverter.py
# coding=utf-8
from __future__ import unicode_literals, division
from MusicXMLReader import MusicXMLReader
import os
import sqlite3

__author__ = 'hsercanatli', 'burakuyar', 'andresferrero', 'sertansenturk'


class ScoreConverter(object):
    _octaves = {"2": ",", "3": "", "4": "\'", "5": "\'\'", "6": "\'\'\'",
                "7": "\'\'\'\'", "r": ""}

    _accidentals = {"-1": "fc", "-4": "fb", "-5": "fk", "-8": "fbm",
                    "1": "c", "4": "b", "5": "k", "8": "bm", "0": ""}

    # notes and accidentals dictionary lilypond
    _notes_western2lily = {"g": "4", "a": "5", "b": "6", "c": "0",
                           "d": "1", "e": "2", "f": "3"}

    _key_sig_accidentals = {'double-slash-flat': "(- BUYUKMUCENNEP)",
                            'flat': "(- KUCUK)",
                            'slash-flat': "(- BAKIYE)",
                            'quarter-flat': "(- KOMA)",
                            'slash-sharp': "BUYUKMUCENNEP",
                            'slash-quarter-sharp': "KUCUK",
                            'sharp': "BAKIYE",
                            'quarter-sharp': "KOMA"}

    _makam_accidentals = {'quarter-flat': '-1',
                          'slash-flat': '-4',
                          'flat': '-5',
                          'double-slash-flat': '-8',
                          'quarter-sharp': '+1',
                          'sharp': '+4',
                          'slash-quarter-sharp': '+5',
                          'slash-sharp': '+8'}

    # sorting rules of key signatures
    sort_rule_sharps = {'F': 0, 'C': 1, 'G': 2, 'D': 3, 'A': 4, 'E': 5,
                        'B': 6}
    sort_rule_notes_sharps = {0: 'F', 1: 'C', 2: 'G', 3: 'D', 4: 'A',
                              5: 'E', 6: 'B'}

    sort_rule_flats = {'F': 6, 'C': 5, 'G': 4, 'D': 3, 'A': 2, 'E': 1,
                       'B': 0}
    sort_rule_notes_flats = {6: 'F', 5: 'C', 4: 'G', 3: 'D', 2: 'A',
                             1: 'E', 0: 'B'}

    db_path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                           "data", "symbtr.db")

    @classmethod
    def _write_lilypond(cls, measures, makam, usul, form, time_sigs,
                        keysig, render_metadata, work_title, composer,
                        lyricist):
        # connecting database, trying to get information for beams in lilypond
        conn = sqlite3.connect(os.path.join(cls.db_path))
        c = conn.cursor()

        # Initialize line number and mapping
        line = 0

        # getting headers
        if render_metadata:
            if lyricist and lyricist != '-':
                poet_str = """\n  poet = \"{0:s}\"""".format(lyricist)
                line += 1
            else:
                poet_str = ''

            metadata_str = """
  title = \"{0}\"
  composer = \"{1}\"
  piece = \"Makam: {2}, Form: {3}, Usul: {4}\"""".format(
                work_title, composer, form, makam, usul) + poet_str
            line += 3
        else:
            metadata_str = ''

        ly_stream = ["""
\\include "makam.ly" """ + """
\\header {
  tagline = \"\"""" + metadata_str +
                     """
}
{
  %\\override Score.SpacingSpanner.strict-note-spacing = ##t
  %\\set Score.proportionalNotationDuration = #(ly:make-moment 1/8)"""]
        line += 8

        # getting the components for the given usul
        c.execute('SELECT * FROM usul WHERE NAME="{0}"'.format(usul))
        data = c.fetchone()
        if data is None:
            c.execute('SELECT * FROM usul WHERE NAMEENG="{0}"'.format(
                usul.lower()))
            data = c.fetchone()

        # if beam information is exist
        if data is not None:
            if data[-1] is not None:
                strokes = data[-1].replace("+", " ")
                tmp_str = '''\n  \\set Staff.beatStructure = #\'({0})''' \
                    .format(strokes)
                ly_stream.append(tmp_str)
                line += 1

        # first time_signature
        ly_stream.append("\n  \\time")
        ly_stream.append("{0:s}/{1:s}".format(time_sigs[0]['beats'],
                                              time_sigs[0]['beat_type']))
        time_sigs.pop(0, None)  # remove the first time signature
        line += 1

        ly_stream.append("\n  \\clef treble")
        ly_stream.append("\n  \\set Staff.keySignature = #`(")
        line += 2

        accidentals_check = []
        rule = []
        for queue in keysig.keys():
            if cls._makam_accidentals[keysig[queue]][0] is "+":
                rule.append([cls.sort_rule_sharps[queue], 's'])
            else:
                rule.append([cls.sort_rule_flats[queue], 'f'])

        # sorting of key signatures
        temp_keysig = ""
        for queue in sorted(rule):
            if queue[1] is "s":
                key = cls.sort_rule_notes_sharps[queue[0]]
            else:
                key = cls.sort_rule_notes_flats[queue[0]]
            accidentals_check.append(key + cls._makam_accidentals[keysig[key].
                                     replace("+", "")])
            temp_keysig += "("
            temp_keysig += (cls._notes_western2lily[key.lower()] + " . ," +
                            cls._key_sig_accidentals[keysig[key]])
            temp_keysig += ") "

            ly_stream.append(temp_keysig)
            temp_keysig = ""

        ly_stream.append(")")

        mapping = []
        for mm, measure in enumerate(measures):
            ly_stream.append("\n  {{ % measure {0:d} beginning".format(mm + 1))
            line += 1

            # apply time signature change, if exists
            if mm in time_sigs.keys():
                # first time_signature
                ly_stream.append("\n    \\time")
                ly_stream.append("{0:s}/{1:s}".format(
                    time_sigs[mm]['beats'], time_sigs[mm]['beat_type']))
                line += 1

            tuplet = 0
            for note in measure:
                temp_note = "\n    "
                line += 1
                if note[6] is None:  # gracenote
                    # for now we display it as a 8th \grace
                    # TODO: Introduce other gracenotes such as \acciaccatura
                    # with the proper symbolic "duration"
                    temp_note += "\\grace "
                    note[6] = 0.5  # reminder: 4th note is 1
                    temp_dur = 4.0 / note[6]  # normal duration
                else:
                    temp_dur = 4.0 / note[6]  # normal duration

                # dotted
                if note[3] == 1:  # dot flag
                    temp_note += note[0]  # step
                    # accidental
                    temp_note += cls._accidentals[
                        str(note[2]).replace('+', '')]
                    temp_note += cls._octaves[note[1]]  # octave

                    temp_dur = temp_dur * 3.0 / 2
                    temp_note += str(int(temp_dur))
                    temp_note += "."
                elif note[4] == 1:  # tuplet flag
                    if tuplet == 0:
                        tuplet = 4
                        temp_note += "\\tuplet 3/2 {\n    "
                        line += 1
                    temp_note += '  '  # increase indentation
                    temp_note += note[0]  # step
                    # accidental
                    temp_note += cls._accidentals[
                        str(note[2]).replace('+', '')]
                    temp_note += cls._octaves[note[1]]  # octave

                    temp_dur = temp_dur * 2.0 / 3
                    temp_note += str(int(temp_dur))

                    tuplet -= 1
                else:  # normal
                    temp_note += note[0]
                    temp_note += cls._accidentals[
                        str(note[2]).replace('+', '')]
                    temp_note += cls._octaves[note[1]]
                    temp_note += str(int(temp_dur))
                if note[7]:
                    pos = len('    ')  # notes start with 4 spaces
                    mapping.append((note[7], pos, line))

                # lyrics
                if note[-1] is not "":
                    if len(note[-1]) > 1:
                        if note[-1][1].isupper() or note[-1][0].isdigit():
                            temp_note += (
                                '''^\\markup { \\left-align ''' +
                                '''{\\bold \\translate #'(1 . 0) \"''' +
                                ''.join(note[-1]).strip() + '''\"}}''')
                        else:
                            temp_note += (
                                '''_\\markup { \\center-align {\\smaller ''' +
                                '''\\translate #'(0 . -2.5) \"''' +
                                ''.join(note[-1]).strip() + '''\"}}''')
                    else:
                        temp_note += (
                            '''_\\markup { \\center-align {\\smaller ''' +
                            '''\\translate #'(0 . -2.5) \"''' +
                            ''.join(note[-1]).strip() + '''\"}}''')

                temp_note += ' % SymbTr-txt row index #' + str(note[7])
                if tuplet == 1:
                    temp_note += "\n    }"
                    tuplet = 0
                    line += 1

                ly_stream.append(temp_note)
            if mm == len(measures) - 1:
                ly_stream.append('''\n    \\bar \"|.\"''')  # closing bar
            ly_stream.append("\n  } % measure " + str(mm + 1) + " ending")
            line += 1

        ly_stream.append("\n}")  # close lilypond

        return ly_stream, mapping

    @classmethod
    def convert(cls, xml_in, ly_out=None, render_metadata=True):
        (measures, makam, usul, form, time_sigs, keysig, work_title,
         composer, poet) = MusicXMLReader.read(xml_in)

        ly_stream, mapping = cls._write_lilypond(
            measures, makam, usul, form, time_sigs, keysig,
            render_metadata, work_title, composer, poet)

        ly_stream = u''.join(ly_stream)
        # save to file
        if ly_out is not None:
            with open(ly_out, 'w') as outfile:
                outfile.write(ly_stream.encode('utf-8'))

        return ly_stream, mapping


musicxml2lilypond/musicxml2lilypond.py
# coding=utf-8
from __future__ import unicode_literals, division
from .MusicXMLReader import MusicXMLReader
import os
import json
import sqlite3

__author__ = 'hsercanatli', 'burakuyar', 'andresferrero', 'sertansenturk'


class ScoreConverter(object):
    @staticmethod
    def lilypond_writer(measures, makam, usul, form, beats, beat_type,
                        keysig, render_metadata, work_title, composer, poet):
        mapping = []

        db_path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                               "data", "symbtr.db")

        # connecting database, trying to get information for beams in lilypond
        conn = sqlite3.connect(os.path.join(db_path))
        c = conn.cursor()

        # getting headers
        if render_metadata:
            poet_str = 'poet = \"Lyricist: {0:s}\"'.\
                format(poet) if poet else ''

            ly_stream = ["""
    \\include "makam.ly" """ + """
    \\header { """ + """
          tagline = \"\"
          title = \"{0}\"
          composer = \"{1}\"
          metre = \"Usul: {2}\"
          piece = \"Makam: {4}, Form: {3}\"""".format(
                work_title, composer, usul, form, makam) + poet_str + "\n}" +
                         """
    {
      %\\override Score.SpacingSpanner.strict-note-spacing = ##t
      %\\set Score.proportionalNotationDuration = #(ly:make-moment 1/8)
                 """]
        else:
            ly_stream = ["""
    \\include "makam.ly" """ + """
    \\header { """ + """
          tagline = \"\"
          """"\n\t\t}" + """
    {
      %\\override Score.SpacingSpanner.strict-note-spacing = ##t
      %\\set Score.proportionalNotationDuration = #(ly:make-moment 1/8)
                 """]

        octaves = {"2": ",", "3": "", "4": "\'", "5": "\'\'", "6": "\'\'\'",
                   "7": "\'\'\'\'", "r": ""}

        accidentals = {"-1": "fc", "-4": "fb", "-5": "fk", "-8": "fbm",
                       "1": "c", "4": "b", "5": "k", "8": "bm", "0": ""}

        # notes and accidentals dictionary lilypond
        notes_western2lily = {"g": "4", "a": "5", "b": "6", "c": "0",
                              "d": "1", "e": "2", "f": "3"}

        notes_keyaccidentals = {'double-slash-flat': "(- BUYUKMUCENNEP)",
                                'flat': "(- KUCUK)",
                                'slash-flat': "(- BAKIYE)",
                                'quarter-flat': "(- KOMA)",
                                'slash-sharp': "BUYUKMUCENNEP",
                                'slash-quarter-sharp': "KUCUK",
                                'sharp': "BAKIYE",
                                'quarter-sharp': "KOMA"}

        makam_accidents = {'quarter-flat': '-1',
                           'slash-flat': '-4',
                           'flat': '-5',
                           'double-slash-flat': '-8',
                           'quarter-sharp': '+1',
                           'sharp': '+4',
                           'slash-quarter-sharp': '+5',
                           'slash-sharp': '+8'}

        # sorting rules of key signatures
        sort_rule_sharps = {'F': 0, 'C': 1, 'G': 2, 'D': 3, 'A': 4, 'E': 5,
                            'B': 6}
        sort_rule_notes_sharps = {0: 'F', 1: 'C', 2: 'G', 3: 'D', 4: 'A',
                                  5: 'E', 6: 'B'}

        sort_rule_flats = {'F': 6, 'C': 5, 'G': 4, 'D': 3, 'A': 2, 'E': 1,
                           'B': 0}
        sort_rule_notes_flats = {6: 'F', 5: 'C', 4: 'G', 3: 'D', 2: 'A',
                                 1: 'E', 0: 'B'}

        # Starting from 4 because of the lilypond header, defined in main func
        line = 6
        # getting the components for the given usul
        c.execute('SELECT * FROM usul WHERE NAME="{0}"'.format(usul))
        data = c.fetchone()

        # if beam information is exist
        if data is not None:
            if data[-1] is not None:
                strokes = data[-1].replace("+", " ")
                tmp_str = '''\n\t\\set Staff.beatStructure = #\'({0})\n'''\
                    .format(strokes)
                ly_stream.append(tmp_str)
                line += 2
        if data is None:
            c.execute('SELECT * FROM usul WHERE NAMEENG="{0}"'.format(
                usul.lower()))
            new_data = c.fetchone()
            if new_data is not None:
                if data[-1] is not None:
                    strokes = new_data[-1].replace("+", " ")
                    tmp_str = '''\n\t\\set Staff.beatStructure = #\'({0})\n'''\
                        .format(strokes)
                    ly_stream.append(tmp_str)
                    line += 2

        ly_stream.append("\n\t\\time")
        line += 1

        # time signature
        ly_stream.append(beats + "/" + beat_type)

        ly_stream.append("\n\t\\clef treble \n\t\\set Staff.keySignature = #`("
                         )
        line += 2

        accidentals_check = []
        rule = []
        for queue in keysig.keys():
            if makam_accidents[keysig[queue]][0] is "+":
                rule.append([sort_rule_sharps[queue], 's'])
            else:
                rule.append([sort_rule_flats[queue], 'f'])

        # sorting of key signatures
        temp_keysig = ""
        for queue in sorted(rule):
            if queue[1] is "s":
                key = sort_rule_notes_sharps[queue[0]]
            else:
                key = sort_rule_notes_flats[queue[0]]
            accidentals_check.append(key + makam_accidents[keysig[key].
                                     replace("+", "")])
            temp_keysig += "("
            temp_keysig += (notes_western2lily[key.lower()] + " . ," +
                            notes_keyaccidentals[keysig[key]])
            temp_keysig += ") "

            ly_stream.append(temp_keysig)
            temp_keysig = ""

        ly_stream.append(")")

        for xx, measure in enumerate(measures):
            ly_stream.append("\n\t")
            line += 1
            ly_stream.append("\n\t{{ % measure {0:d} beginning".format(xx + 1))

            tuplet = 0
            pos = 0

            for note in measure:
                temp_note = "\n\t"
                line += 1
                if note[6] is None:  # gracenote
                    # for now we display it as a 8th \grace
                    # TODO: Introduce other gracenotes such as \acciaccatura
                    # with the proper symbolic "duration"
                    temp_note += "\\grace "
                    note[6] = 0.5  # reminder: 4th note is 1
                    temp_dur = 4.0 / note[6]  # normal duration
                else:
                    temp_dur = 4.0 / note[6]  # normal duration

                # dotted
                if note[3] == 1:  # dot flag
                    temp_note += note[0]  # step
                    # accidental
                    temp_note += accidentals[str(note[2]).replace('+', '')]
                    temp_note += octaves[note[1]]  # octave

                    temp_dur = temp_dur * 3.0 / 2
                    temp_note += str(int(temp_dur))
                    temp_note += "."
                elif note[4] == 1:  # tuplet flag
                    if tuplet == 0:
                        tuplet = 4
                        temp_note += "\\tuplet 3/2 {\n\t"
                    temp_note += note[0]  # step
                    # accidental
                    temp_note += accidentals[str(note[2]).replace('+', '')]
                    temp_note += octaves[note[1]]  # octave

                    temp_dur = temp_dur * 2.0 / 3
                    temp_note += str(int(temp_dur))

                    tuplet -= 1
                else:  # normal
                    temp_note += note[0]
                    temp_note += accidentals[str(note[2]).replace('+', '')]
                    temp_note += octaves[note[1]]
                    temp_note += str(int(temp_dur))
                if note[7]:
                    mapping.append((note[7], pos + 4, line))

                # lyrics
                if note[-1] is not "":
                    if len(note[-1]) > 1:
                        if note[-1][1].isupper() or note[-1][0].isdigit():
                            temp_note += (
                                '''^\\markup { \\left-align ''' +
                                '''{\\bold \\translate #'(1 . 0) \"''' +
                                ''.join(note[-1]).strip() + '''\"}}''')
                        else:
                            temp_note += (
                                '''_\\markup { \\center-align {\\smaller ''' +
                                '''\\translate #'(0 . -2.5) \"''' +
                                ''.join(note[-1]).strip() + '''\"}}''')
                    else:
                        temp_note += (
                            '''_\\markup { \\center-align {\\smaller ''' +
                            '''\\translate #'(0 . -2.5) \"''' +
                            ''.join(note[-1]).strip() + '''\"}}''')

                if tuplet == 1:
                    temp_note += "\n\t }"
                    tuplet = 0

                temp_note += ' % SymbTr-txt #' + str(note[7])
                pos += len(temp_note) + 1
                ly_stream.append(temp_note)

            ly_stream.append("\n\t} % measure " + str(xx + 1) + " end point")
        ly_stream.append('''\n\t\\bar \"|.\"''' + "\n}")
        return ly_stream, mapping

    @classmethod
    def run(cls, xml_in, ly_out=None, map_out=None, render_metadata=False):
        (measures, makam, usul, form, beats, beat_type, keysig, work_title,
         composer, poet) = MusicXMLReader.read(xml_in)

        ly_stream, mapping = cls.lilypond_writer(
            measures, makam, usul, form, beats, beat_type, keysig,
            render_metadata, work_title, composer, poet)

        ly_stream = u''.join(ly_stream)
        # save to file
        if ly_out is not None:
            xml_in = xml_in.split(".")[0]
            with open(ly_out, 'w') as outfile:
                outfile.write(ly_stream.encode('utf-8'))

        # save to json
        if map_out is not None:
            with open(xml_in + ".json", 'w') as outfile:
                json.dump(mapping, outfile)

        return ly_stream, mapping
